src/daft-logical-plan/src/optimization/rules/lift_project_from_agg.rs

```diff
@@ -0,0 +1,264 @@
+use std::sync::Arc;
+
+use common_error::DaftResult;
+use common_treenode::{Transformed, TreeNode};
+use daft_dsl::{col, optimization::requires_computation, Expr};
+use indexmap::IndexSet;
+
+use super::OptimizerRule;
+use crate::{
+    ops::{Aggregate, Project},
+    LogicalPlan,
+};
+
+/// Optimization rule for lifting expressions that can be done in a project out of an aggregation.
+///
+/// After a pass of this rule, the top level expressions in each aggregate should all be aliases or agg exprs.
```

Comment:
@kevinzwang, so i know this is just semantics, but this seems like more of a rewrite instead of an optimization. I wonder if it's worth making the distinction between optimizations and rewriting.

Reply:
Yeah that's true. I mean technically I have this rule because moving things out into a project allows for other optimizations, however the implementation of the aggregate translation assumes that this rule is applied. 

In general, should it be fine to have rewriting rules in the optimizer instead of in each of the ops themselves? AKA is it okay for the logical plan to be in an invalid state until it has been optimized? I personally don't have very strong thoughts about this

Reply:
🤔 I don't like the idea of having invalid plans, as that wouldnt allow us to disable certain optimizations.

I know there's been some discussion with @samster25 about potentially introducing a 'unresolved' or 'dsl' plan that wouldn't be guaranteed to be semantically valid, only syntactically valid. Maybe just add a comment [here](https://github.com/Eventual-Inc/Daft/pull/3296/files#diff-63dbdc903bf0e71ab768f3fb54ae57712e7efaa14bbd90d867ae326b8ac273f7R109) saying that this is actually a 'rewrite' to make it into a valid plan, just so we don't lose track of it over time.